### PR TITLE
RATIS-2164. LeakDetector has a race condition.

### PR DESCRIPTION
**Pull request by szetszwo**
RATIS-2164

This is mainly fixing the first two problem mentioned in https://github.com/apache/ratis/pull/1156

1. `LeakDetector`: asserted allLeaks is non-empty but printed "allLeaks.size = 0"
    - Another bug: Trackers can be considered as a leak without calling `retain()`. Without calling `retain()` at all, it is not a leak.
2. `SimpleTracing` and `AdvancedTracing`: the methods should be synchronized.
    - Minor presentation problem: `AdvancedTracing` should have a single track list instead of two lists: `retainsTraces` and `releaseTraces`.

### CODE REVIEW

**Comment by szetszwo** [comment]
@duongkame , @133tosakarin , @OneSizeFitsQuorum , @adoroszlai , please take a look.

**Comment by szetszwo** [comment]
The failures are expected since it needs the other fixes described in https://github.com/apache/ratis/pull/1156

**Review by 133tosakarin — APPROVED** [review]

**Comment by szetszwo** [comment]
@OneSizeFitsQuorum , @adoroszlai , we need a committer review approval for this.  With this and RATIS-2173, the tests will be much easier to pass (although there may be still some undiscovered bugs.)

**Review by adoroszlai — APPROVED** [review]

**Comment by szetszwo** [comment]
@adoroszlai , thanks a lot for reviewing this!

Although the test failures are unrelated, let's rerun them one more time.

**Pull request merged**
